internal/pod_helper.go

```diff
@@ -127,44 +130,78 @@ func GetJSONHash(object interface{}) (string, error) {
 
 // GetPodLabels creates the labels that we will apply to a Pod
 func GetPodLabels(cluster *fdbtypes.FoundationDBCluster, processClass fdbtypes.ProcessClass, id string) map[string]string {
-	labels := map[string]string{}
+	matchLabels := map[string]string{}
 
 	for key, value := range cluster.Spec.LabelConfig.MatchLabels {
-		labels[key] = value
+		matchLabels[key] = value
 	}
 
 	if processClass != "" {
 		for _, label := range cluster.Spec.LabelConfig.ProcessClassLabels {
-			labels[label] = string(processClass)
+			matchLabels[label] = string(processClass)
 		}
 	}
 
 	if id != "" {
 		for _, label := range cluster.Spec.LabelConfig.ProcessGroupIDLabels {
-			labels[label] = id
+			matchLabels[label] = id
 		}
 	}
 
-	return labels
+	return matchLabels
 }
 
 // GetPodMatchLabels creates the labels that we will use when filtering for a pod.
-func GetPodMatchLabels(cluster *fdbtypes.FoundationDBCluster, processClass fdbtypes.ProcessClass, id string) map[string]string {
-	labels := map[string]string{}
+func GetPodMatchLabels(cluster *fdbtypes.FoundationDBCluster, processClass fdbtypes.ProcessClass, processGroupID string) map[string]string {
+	matchLabels := map[string]string{}
 
 	for key, value := range cluster.Spec.LabelConfig.MatchLabels {
-		labels[key] = value
+		matchLabels[key] = value
 	}
 
 	if processClass != "" {
-		labels[cluster.GetProcessClassLabel()] = string(processClass)
+		matchLabels[cluster.GetProcessClassLabel()] = string(processClass)
 	}
 
-	if id != "" {
-		labels[cluster.GetProcessGroupIDLabel()] = id
+	if processGroupID != "" {
+		matchLabels[cluster.GetProcessGroupIDLabel()] = processGroupID
+	}
+
+	return matchLabels
+}
+
+func generateSelector(cluster *fdbtypes.FoundationDBCluster, processClass fdbtypes.ProcessClass, processGroupID string) labels.Selector {
+	matchLabels := GetPodMatchLabels(cluster, processClass, processGroupID)
+	s := labels.SelectorFromSet(matchLabels)
+
+	// If the process class is not specified we still want to ensure that the Pod has
+	// a label with the that key.
+	if _, ok := matchLabels[cluster.GetProcessClassLabel()]; !ok {
```

Comment:
This feels like it's accomplishing something indirect, and I wonder if it is better to achieve this by the user setting the match labels differently.

Reply:
Since the `GetPodMatchLabels` returns a map I'm not sure if we could do something better here but I also don't see an issue with the current approach. The problem would be if we add the according label with an empty value the selector will expect an empty value too.

Reply:
Sounds like a valid approach to me 👍